### ISSUE TITLE
thttpd has large memory footprint

### ISSUE THREAD

**Issue by wc7086** [vergoh/vnstat-docker#1]
I have observed that thttpd has a memory footprint of up to 148MiB. What are the advantages of this program?

**Comment by wc7086** [comment]
Is it because of this?
https://marc.info/?l=thttpd&m=102182637422558&w=2

**Comment by vergoh** [comment]
I haven't actually checked how much thttpd is using memory and how it would possibly compare to other alternatives. Thanks for bringing up this as a possible concern.

The reason why I ended up going with thttpd was because it was the smallest install footprint httpd having built-in cgi support that I could find from the Alpine repository. It's also simple from configuration point of view which fits rather well the needs of this usage and not currently considered as discontinued. Althttpd was also one candidate but it felt a little bit too much hardcoded for serving sqlite.org content, some user-agents blocks are for example directly fixed in the source code.

**Comment by vergoh** [comment]
I'm not really sure what's going on with the thttpd memory usage now that I look at it. It does indeed appear to have a high footprint based on `top` but also `pmap` output even right after being started without any requests being done:

```
$ sudo pmap 3351
3351:   thttpd -C /etc/thttpd.conf -p 8685 -l /dev/stdout
000055e8f7418000     20K r---- thttpd
000055e8f741d000     52K r-x-- thttpd
000055e8f742a000     24K r---- thttpd
000055e8f7430000      4K r---- thttpd
000055e8f7431000      8K rw--- thttpd
000055e8f7433000      4K rw---   [ anon ]
000055e8f8cd5000      4K -----   [ anon ]
000055e8f8cd6000      4K rw---   [ anon ]
00007f61455b2000 176148K rw---   [ anon ]
00007f61501b7000      4K rw---   [ anon ]
00007f61501b8000      4K r--s- localtime
00007f61501b9000     84K r---- ld-musl-x86_64.so.1
00007f61501ce000    288K r-x-- ld-musl-x86_64.so.1
00007f6150216000    216K r---- ld-musl-x86_64.so.1
00007f615024c000      4K r---- ld-musl-x86_64.so.1
00007f615024d000      4K rw--- ld-musl-x86_64.so.1
00007f615024e000     12K rw---   [ anon ]
00007ffdede6b000    132K rw---   [ stack ]
00007ffdedfa6000     12K r----   [ anon ]
00007ffdedfa9000      4K r-x--   [ anon ]
ffffffffff600000      4K --x--   [ anon ]
 total           177036K
```

There are few configuration options available in config.h (`DESIRED_MAX_MAPPED_FILES` and `DESIRED_MAX_MAPPED_BYTES`) which should affect the memory usage while the output files `vnstati` generates are less than 1 MiB total even for several interfaces.

I tested compiling thttpd from scratch with those settings minimized but those don't appear to result in any change to the memory usage for some reason:

```
$ sudo pmap 4972
4972:   ./thttpd -C /etc/thttpd.conf -p 8685 -l /dev/stdout
000055f07abbe000     20K r---- thttpd
000055f07abc3000     52K r-x-- thttpd
000055f07abd0000     24K r---- thttpd
000055f07abd6000      4K r---- thttpd
000055f07abd7000      8K rw--- thttpd
000055f07abd9000      4K rw---   [ anon ]
000055f07b6ed000      4K -----   [ anon ]
000055f07b6ee000      4K rw---   [ anon ]
00007f5796c42000 176148K rw---   [ anon ]
00007f57a1847000      4K rw---   [ anon ]
00007f57a1848000      4K r--s- localtime
00007f57a1849000     84K r---- ld-musl-x86_64.so.1
00007f57a185e000    288K r-x-- ld-musl-x86_64.so.1
00007f57a18a6000    216K r---- ld-musl-x86_64.so.1
00007f57a18dc000      4K r---- ld-musl-x86_64.so.1
00007f57a18dd000      4K rw--- ld-musl-x86_64.so.1
00007f57a18de000     12K rw---   [ anon ]
00007ffee9462000    132K rw---   [ stack ]
00007ffee9503000     12K r----   [ anon ]
00007ffee9506000      4K r-x--   [ anon ]
ffffffffff600000      4K --x--   [ anon ]
 total           177036K
```

**Comment by vergoh** [comment]
I spent several hours trying to understand what's going on with the thttpd memory usage but didn't find any good answers. With all the debugs enabled, thttpd shows that it isn't allocating that much memory at least for the hash or cache initially, or it's not aware of it. There's also no obvious memory allocations that I could see in the implementation that could explain those numbers but that output suggests the allocation still exists somewhere for some reason.

Eventually I concluded that I'm going to replace thttpd with something else as you aren't likely going to be the only one with this memory usage question and the current status doesn't indeed look right. I had one implementation running with BusyBox httpd, and while it was tiny from resource usage point of view, it didn't have a good log output (no timestamps, requester IP and user-agent missing etc.) and didn't support serving cgis from the server root directory. As a result, the current plan is to replace thttpd with lighttpd.
